crates/editor/src/editor.rs

```diff
@@ -7287,17 +7287,58 @@ impl Editor {
         let buffer = self.buffer.read(cx);
         let head = self.selections.newest::<usize>(cx).head();
         let (buffer, head) = buffer.text_anchor_for_position(head, cx)?;
+        let head_offset = text::ToOffset::to_offset(&head, buffer.read(cx));
         let replica_id = self.replica_id(cx);
 
         let workspace = self.workspace()?;
         let project = workspace.read(cx).project().clone();
         let references = project.update(cx, |project, cx| project.references(&buffer, head, cx));
-        Some(cx.spawn(|_, mut cx| async move {
-            let locations = references.await?;
+        Some(cx.spawn(|editor, mut cx| async move {
+            let mut locations = references.await?;
+
+            // Lsp may return references that contain self pos (eg. rust-analyzer)
+            // So we remove self pos from locations
+            editor.update(&mut cx, |_, cx| {
+                locations = locations
+                    .iter()
+                    .filter(|r| !r.range.to_offset(r.buffer.read(cx)).contains(&head_offset))
+                    .map(|x| x.clone())
+                    .collect();
+            })?;
+
             if locations.is_empty() {
                 return Ok(());
             }
 
+            // If there is one reference, just open it directly
+            if locations.len() == 1 {
+                let target = locations.pop().unwrap();
+
+                editor.update(&mut cx, |editor, cx| {
+                    let range = target.range.to_offset(target.buffer.read(cx));
+                    let range = editor.range_for_match(&range);
+
+                    if Some(&target.buffer) == editor.buffer().read(cx).as_singleton().as_ref() {
+                        editor.change_selections(Some(Autoscroll::fit()), cx, |s| {
+                            s.select_ranges([range]);
+                        });
+                    } else {
+                        cx.window_context().defer(move |cx| {
+                            let target_editor: View<Self> =
+                                workspace.update(cx, |workspace, cx| {
+                                    workspace.open_project_item(target.buffer.clone(), cx)
+                                });
+                            target_editor.update(cx, |target_editor, cx| {
+                                target_editor.change_selections(Some(Autoscroll::fit()), cx, |s| {
+                                    s.select_ranges([range]);
+                                })
+                            })
+                        })
+                    }
+                })?;
+                return Ok(());
```

Comment:
You can return the value of `editor.update`, right?

Reply:
Yes!